README.md
# fuzzymerge_parallel

Merge two pandas dataframes by using a function to calculate the edit distance (Levenshtein Distance) using multiprocessing for parallelization on a single node or Dask for distributed computation across multiple nodes.

**Efficient Matching and Merging**

Matching and merging data can be demanding in terms of time and memory usage. That's why our package is specifically crafted to address these challenges effectively.

**Optimized Execution**

To boost execution performance, we've fine-tuned our package to work seamlessly in both single-node and multi-node environments. Whether you're processing data on a local machine or distributing tasks across a cluster, our package is optimized to get the job done efficiently.

**Smart Memory Management**

Memory efficiency is a priority. Our algorithm estimates memory requirements and divides the workload into manageable batches. This ensures that your data operations fit comfortably within your available memory. Plus, you have the flexibility to customize these settings to match your specific needs.

With this package, you can confidently tackle data matching and merging tasks while optimizing both time and memory resources.

## Description

fuzzymerge_parallel offers two modes for faster execution:

### 1. Multiprocessing mode

This mode runs on a single machine and it is able to use multi-CPU cores. This mode does not require dask and it's ideal for local processing tasks, utilizing Numpy and Python's multiprocessing libraries to speed things up.

### 2. Dask mode

In this mode, fuzzymerge_parallel utilizes a Dask client, which can be configured for single or multi-node setups. To leverage the dask mode it is suggested to use it multiple nodes. Multi-node dask clients distribute computations across clusters of machines, making it suitable for heavy-duty processing. Using dask mode offers numerous benefits, automating tasks that would otherwise require manual intervention, such as enhancing performance, expanding scalability, ensuring fault tolerance, optimizing resource utilization, and enabling parallelism.

### Features

- Performs fuzzy merging of dataframes based on string columns
- Utilize distance functions (e.g., Levenshtein) for intelligent matching
- Benefit from parallel computing techniques for enhanced performance
- Easily integrate into your existing data processing pipelines

## Installation

To download and install the fuzzymerge_parallel Python package from GitHub, you can follow these improved instructions:

To install FuzzyMergeParallel via pip from its GitHub repository, follow these steps:

1. **Download the Package:** Begin by downloading the package from GitHub. You can use git to clone the repository to your local machine:
    ```bash
    git clone https://github.com/ULHPC/fuzzymerge_parallel.git
    ```
    
2. **Navigate to the Package Directory:** Open a terminal or command prompt and change your current directory to the downloaded package folder:
    ```bash
    cd fuzzymerge_parallel
    ```

3. **Install the Package:** Finally, use pip to install the package in "editable" mode (with the -e flag) to allow for development and updates. There are two options:

    **Option 1:** If you plan to use the package on a single node and don't need the dask and distributed dependencies, simply run:
    ```bash
    pip install -e .
    ```

    **Option 2:** If you intend to use the package in both single and multi-node environments with dask and distributed support, use the following command:
    ```bash
    pip install -e ".[dask]"
    ```


This command will install the package along with its dependencies. You can now import and use FuzzyMergeParallel in your Python projects.    

## Dependencies

To use this package, you will need to have the following dependencies installed:

- [Click](https://pypi.org/project/Click/) >= 7.0
- [dask[distributed]](https://pypi.org/project/dask/) >= 2023.5.0 (Optional: Only needed for multi-node)
- [Levenshtein](https://pypi.org/project/python-Levenshtein/) >= 0.21.0 (Optional: Only needed for multi-node)
- [nltk](https://pypi.org/project/nltk/) >= 3.8.1
- [numpy](https://pypi.org/project/numpy/) >= 1.23.5
- [pandas](https://pypi.org/project/pandas/) >= 1.5.3
- [tqdm](https://pypi.org/project/tqdm/) >= 4.65.0
- [psutil](https://pypi.org/project/psutil/) == 5.9.5
- [pytest](https://pypi.org/project/pytest/) >= 7.4.1

## Description

The FuzzyMergeParallel class is exposed and it is highly configurable. The following parameters and other attributes can be set up before doing the merge operation:



| Parameter        | Description                                                      |
|------------------|------------------------------------------------------------------|
| left             | The left input data to be merged.                                |
| right            | The right input data to be merged.                               |
| left_on          | Column(s) in the left DataFrame to use as merge keys.            |
| right_on         | Column(s) in the right DataFrame to use as merge keys.           |

Example create a FuzzyMergeParallel class:

```python
fuzzy_merger = FuzzyMergeParallel(left_df, right_df, left_on='left_column_name', right_on='right_column_name')
```


| Attribute        | Description                                                      |
|------------------|------------------------------------------------------------------|
| uselower         | Whether to convert strings to lowercase before comparison. Default is True.        |
| threshold        | The threshold value for fuzzy matching similarity. Default is 0.9.                |
| how              | The type of merge to be performed. Default is 'outer'.                               |
| on               | Column(s) to merge on if not specified in left_on or right_on. Default is False.    |
| left_index       | Whether to use the left DataFrame's index as merge key(s). Default is False.       |
| right_index      | Whether to use the right DataFrame's index as merge key(s). Default is False.      |
| parallel         | Whether to perform the merge operation in parallel. Default is True.              |
| n_threads        | The number of threads to use for parallel execution. Default is 'all' (a thread per each available core).             |
| hide_progress    | Whether to display a progress bar during the merge operation. Default is False.    |
| num_batches      | The number of batches to split the ratio computation. Default is automatic.              |
| ratio_function   | The distance ratio function.                Defaults to `Levenshtein.ratio()`.                      |
| dask_client      | A dask client object.                                            |

Example set extra attributes by stating the name of the attribute and its value with `set_parameter()`:

```python
fuzzy_merger.set_parameter('how', 'inner')
fuzzy_merger.set_parameter('threshold', 0.75)
```

## Usage


### Single node 

#### Sequential execution

```python
fuzzy_merger = FuzzyMergeParallel(left_df, right_df, left_on='left_column_name', right_on='right_column_name')
# Set parameters
fuzzy_merger.set_parameter('how', 'inner')
fuzzy_merger.set_parameter('parallel', False)
# Run the merge sequentially
result = fuzzy_merger.merge()
```

#### Multiprocessing execution

```python
fuzzy_merger = FuzzyMergeParallel(left_df, right_df, left_on='left_column_name', right_on='right_column_name')
# Set parameters
fuzzy_merger.set_parameter('how', 'inner')
fuzzy_merger.set_parameter('n_threads', 64)
# Run the merge multiprocessing
result = fuzzy_merger.merge()
```

### Multi-node (dask)

#### Local client

```python
fuzzy_merger = FuzzyMergeParallel(left_df, right_df, left_on='left_column_name', right_on='right_column_name')
# Set parameters
fuzzy_merger.set_parameter('how', 'inner')

# Set parameters for dask
## Create a dask client
from dask.distributed import Client
client = Client(...)  # Connect to distributed cluster and override default
fuzzy_merger.set_parameter('parallel', True)
fuzzy_merger.set_parameter('dask_client', client)
# Run the merge in dask
result = fuzzy_merger.merge()
```

How to create a dask client?

There are different options to create a dask client. Extensive documentation can be found on their websites:

- [General dask documentation](https://docs.dask.org/en/stable/)
- [dask client documentation](https://distributed.dask.org/en/stable/client.html)
- [Dask jobqueue documentation (distributed)](https://jobqueue.dask.org/en/latest/index.html)

A couple of examples:

```python
# Launch dask on a local cluster (singlenode)
from dask.distributed import Client, LocalCluster
# Create a local Dask cluster
cluster = LocalCluster()
# Create a Dask client to connect to the cluster
client = Client(cluster)
```

```python
# Launch dask on a SLURM cluster
from dask_jobqueue import SLURMCluster

cluster = SLURMCluster(
    queue='regular',
    account="myaccount",
    cores=128,
    memory="500 GB"
)

cluster.scale(jobs=10)  # ask for 10 jobs

client = Client(cluster)
```

## Contributing

Contributions are welcome! If you encounter any issues, have suggestions, or want to contribute improvements, please submit a pull request or open an issue on the GitHub repository.


## Authors

- Oscar J. Castro Lopez (oscar.castro@uni.lu)
  - Parallel Computing & Optimisation Group (PCOG) - **University of Luxembourg**


This package is based on the levenpandas package (https://github.com/fangzhou-xie/levenpandas).

## License

This project is licensed under the MIT License.

HISTORY.rst
=======
History
=======

1.0.0 (2023-12-06) - First release on PyPI.

0.1.0 (2023-06-21)

------------------




tests/__init__.py
"""Unit test package for fuzzymerge_parallel."""


tests/test_fuzzymerge_parallel.py
#!/usr/bin/env python

"""Tests for `fuzzymerge_parallel` package."""

import pytest
from click.testing import CliRunner
# from nltk.corpus import movie_reviews
# import nltk
import pandas as pd
from fuzzymerge_parallel.FuzzyMergeParallel import FuzzyMergeParallel
from fuzzymerge_parallel import cli


def dask_is_installed() -> bool:
    """Checks if dask is installed. This is an auxiliary function to skip dask tests if it's not installed.

    Returns:
        bool: True if dask was imported succesfully, Otherwise False.
    """
    try:
        import dask
        library_installed = True
    except ImportError:
        library_installed = False
    return library_installed


pytest.mark.dask = pytest.mark.skipif(
    (not dask_is_installed()),
    reason="Test requires 'dask' extra to be installed."
)


@pytest.fixture(scope="module")
def dataset():
    """Loads a preprocessed subset of the NLTK corpora/movie_reviews dataset.

    This function loads a curated subset of 5000 rows from the corpora/movie_reviews dataset available in the NLTK package. The original dataset, which contains movie-review text data, was processed to create this subset for testing purposes.

    Dataset Source: [NLTK - Movie Reviews Corpus](http://www.cs.cornell.edu/people/pabo/movie-review-data/)

    The subset dataset has already been carefully filtered and tested with levenshtein distance for each pair, resulting in 600 matching records. Therefore, the expected shape of the yielded dataframes is (600, 2).

    Yields:
        pd.DataFrame: Two dataframes, each containing a list of words from preprocessed movie-review data.
    """
    words_left = pd.read_csv('tests/words_left.zip')
    words_right = pd.read_csv('tests/words_right.zip')
    yield words_left, words_right


def test_command_line_interface():
    """Test the CLI."""
    runner = CliRunner()
    result = runner.invoke(cli.main)
    assert result.exit_code == 0
    assert 'fuzzymerge_parallel.cli.main' in result.output
    help_result = runner.invoke(cli.main, ['--help'])
    assert help_result.exit_code == 0
    assert '--help  Show this message and exit.' in help_result.output


def test_sequential(dataset):
    """Basic sequential text execution of FuzzyMergeParallel.
    """
    left_df, right_df = dataset
    fm_seq = FuzzyMergeParallel(
        left_df, right_df, left_on='words_left', right_on='words_right')
    # Set parameters
    fm_seq.set_parameter('how', 'inner')
    fm_seq.set_parameter('parallel', False)
    # Run the merge sequentially
    result = fm_seq.merge()
    assert result.shape == (600, 2)


def test_basic_multiprocessing(dataset):
    """Basic multiprocessing test execution of FuzzyMergeParallel.
    """
    left_df, right_df = dataset
    fm_multi = FuzzyMergeParallel(
        left_df, right_df, left_on='words_left', right_on='words_right')
    # Set parameters
    fm_multi.set_parameter('how', 'inner')
    # Set parameters for multiprocessing
    fm_multi.set_parameter('n_threads', 'all')  # All available cores
    # Run the merge multiprocessing
    result = fm_multi.merge()
    assert result.shape == (600, 2)


@pytest.mark.dask
def test_basic_dask(dataset):
    """Basic dask test execution of FuzzyMergeParallel.
    """
    from dask.distributed import Client, LocalCluster
    left_df, right_df = dataset
    # Create a local Dask cluster
    cluster = LocalCluster()
    # Create a Dask client to connect to the cluster
    client = Client(cluster)
    # Get the total number of cores available in the cluster
    total_cores = sum(client.ncores().values())
    print("total cores ", total_cores)
    fm_dask = FuzzyMergeParallel(
        left_df, right_df, left_on='words_left', right_on='words_right')
    # Set parameters
    fm_dask.set_parameter('how', 'inner')
    # Set parameters for multiprocessing
    fm_dask.set_parameter('parallel', True)
    fm_dask.set_parameter('dask_client', client)
    # Run the merge in dask
    result = fm_dask.merge()
    client.close()
    assert result.shape == (600, 2)


fuzzymerge_parallel/aux.py
"""

Auxiliary functions for FuzzyMergeParallel.

Author: Oscar J. CASTRO-LOPEZ
Date: 2023-06-19

"""
import psutil


def _conditional_print(msg: str, enabled: bool = False):
    if enabled:
        print(msg)


def suggest_batch_number(df1, df2, scenario: int = 5, factor: float = 0.001, print_enabled: bool = False) -> int:
    """Compute how much memory would be required to execute FuzzyMergeParallel between two dataframes.

    Generates scenarios and returns how many rows per batch fit into memory to execute FuzzyMergeParallel.

    Args:
        df1 (pd.DataFrame): A pandas data frame.
        df2 (pd.DataFrame): A pandas data frame.
        scenario (int, optional): Scenario options from worst to best (1 to 8). Defaults to 5.
        factor (float, optional): Custom factor scenario. Defaults to 0.001.
        print_output (bool, optional): Prints the logic of the scenarios. Defaults to False.

    Returns:
        int: Number of rows to process per batch.
    """
    # Calculate the memory usage of each DataFrame in MB
    memory_usage_df1 = df1.memory_usage(deep=True).sum() / (1024 * 1024 * 1024)
    memory_usage_df2 = df2.memory_usage(deep=True).sum() / (1024 * 1024 * 1024)

    # Print the memory usage of each DataFrame
    _conditional_print(
        f"DataFrame 1 memory usage: {round(memory_usage_df1, 2)} GB - Rows: {df1.shape[0]}", print_enabled)
    _conditional_print(
        f"DataFrame 2 memory usage: {round(memory_usage_df2, 2)} GB - Rows: {df2.shape[0]}", print_enabled)

    # Calculate the memory required for the cross product
    # Worst case
    memory_required = (memory_usage_df1 *
                       df2.shape[0] + memory_usage_df2 * df1.shape[0])

    # Get the available system memory
    system_memory = psutil.virtual_memory().available / (1024 * 1024 * 1024)

    # Dictionary of scenarios
    scenarios = {
        1: {'msg': '1. Worst-case all rows match (Non-realistic)', 'mem_req': round(memory_required, 2)},
        2: {'msg': '2. Half rows match (Non-realistic)', 'mem_req':  round(memory_required * 0.5, 2)},
        3: {'msg': '3. Quarter rows match (25%)  (Unlikely)', 'mem_req':  round(memory_required * 0.25, 2)},
        4: {'msg': '4. Ten percent rows match (10%)', 'mem_req':  round(memory_required * 0.1, 2)},
        5: {'msg': '5. Five percent rows match (5%)', 'mem_req':  round(memory_required * 0.05, 2)},
        6: {'msg': '6. 2.5% rows match', 'mem_req':  round(memory_required * 0.025, 2)},
        7: {'msg': '7. 1% rows match', 'mem_req':  round(memory_required * 0.01, 2)},
        8: {'msg': '8. Custom factor', 'mem_req':  round(memory_required * factor, 2)},
    }

    for _, inner_dict in scenarios.items():
        _conditional_print(
            f"Memory required for {inner_dict['msg']}: {inner_dict['mem_req']} GB", print_enabled)

    # Compare the memory required with the available system memory
    if memory_required <= system_memory:
        _conditional_print(
            "Available system memory is sufficient.", print_enabled)
    else:
        _conditional_print(
            "Insufficient system memory for the cross product operation.", print_enabled)

    _conditional_print(
        f"You selected: {scenarios[scenario]['msg']}", print_enabled)
    memory_required = scenarios[scenario]['mem_req']

    overhead = 0.3
    available_memory = system_memory * (1 - overhead)
    _conditional_print(
        f"Avaliable memory - overhead: {round(available_memory, 2)}", print_enabled)
    batch_size = memory_required / available_memory
    batch_rows = df1.shape[0] / batch_size
    if batch_rows > df1.shape[0]:
        batch_rows = df1.shape[0]
    _conditional_print(
        f"Rows per batch: {round(batch_rows, 2)}", print_enabled)
    return int(batch_rows)


fuzzymerge_parallel/FuzzyMergeParallel.py
"""

FuzzyMergeParallel.

Package for performing fuzzy merging of two dataframes based on a string column, using a distance function such as Levenshtein.

FuzzyMergeParallel offers two modes for faster execution:

1. Multiprocessing mode: This mode runs on a single machine using multiple CPU cores. It's ideal for local processing tasks, utilizing Numpy and Python's multiprocessing libraries to speed things up.
2. Dask mode: In this mode, FuzzyMergeparallel utilizes a Dask client, which can be configured for single or multi-node setups. Multi-node clients distribute computations across clusters of machines, making it suitable for heavy-duty processing.


Author: Oscar J. CASTRO-LOPEZ
Date: 2023-06-19

Classes:
    FuzzyMergeParallel:
        A class for configuring and executing parallel fuzzy merging.

Functions:
    FuzzyMergeParallel.merge(args):
        Executes the fuzzy merge operation.

Notes:
    The code can also run on a single procesor.
    This class is based on the Levenpandas package: https://github.com/fangzhou-xie/levenpandas
"""
from typing import Any
import numpy as np
import pandas as pd
from fuzzymerge_parallel.aux import suggest_batch_number

from fuzzymerge_parallel.matches_multiproc import match_by_left_multiproc
from Levenshtein import ratio as levenshtein_ratio
from tqdm import tqdm


class FuzzyMergeParallel:
    """A class for performing fuzzy merges between two Pandas data frames.

    The merge can be executed sequentially, in parallel using multiprocessing (single-node), or in parallel using Dask (multi-node).

    Usage:
    1. Initialize an instance of the FuzzyMergeParallel class with the input Pandas data frames and the columns to merge.
    2. Configure the matching criteria, threshold values, computing options, and other parameters for the merge operation.
    3. Invoke the merge method to perform the fuzzy merge between the data frames.
    4. Access the merged data frame or retrieve summary statistics about the matches.

    If using Dask, a Dask client object must be provided.

    Examples:
    ```
    fuzzy_merger = FuzzyMergeParallel(left_df, right_df, left_on='left_column_name', right_on='right_column_name')
    # Set parameters
    fuzzy_merger.set_parameter('how', 'inner')
    fuzzy_merger.set_parameter('parallel', False)
    # Run the merge sequentially
    result = fuzzy_merger.merge()

    # Set parameters for multiprocessing
    
    fuzzy_merger.set_parameter('n_threads', 64)
    # Run the merge multiprocessing
    result = fuzzy_merger.merge()
    # Set parameters for dask
    ## Create a dask client
    from dask.distributed import Client
    client = Client(...)  # Connect to distributed cluster and override default
    fuzzy_merger.set_parameter('parallel', True)
    fuzzy_merger.set_parameter('dask_client', client)
    # Run the merge in dask
    result = fuzzy_merger.merge()
    ```
    """

    def __init__(self, left, right, left_on=None, right_on=None):
        """
        Initialize a FuzzyMergeParallel object.

        Parameters:
            left: DataFrame
                The left input data to be merged.
            right: DataFrame
                The right input data to be merged.
            left_on: str, optional
                Column(s) in the left DataFrame to use as merge keys.
            right_on: str, optional
                Column(s) in the right DataFrame to use as merge keys.


        Other Attributes:
            uselower: bool
                Whether to convert strings to lowercase before comparison. Default is True.
            threshold: float
                The threshold value for fuzzy matching similarity. Default is 0.9.
            how: str, optional
                The type of merge to be performed. Default is 'outer'.
            on: str or None
                Column(s) to merge on if not specified in left_on or right_on.
            left_index: bool
                Whether to use the left DataFrame's index as merge key(s). Default is False.
            right_index: bool
                Whether to use the right DataFrame's index as merge key(s). Default is False.
            parallel: bool
                Whether to perform the merge operation in parallel. Default is True.
            n_threads: int
                The number of threads to use for parallel execution. Default is 'all' which runs one thread per available core.
            hide_progress: bool
                Whether to display a progress bar during the merge operation. Default is False.
            num_batches: int
                The number of batches to split the ratio computation. Default is automatic.
            ratio_function: runnable
                The distance ratio function. Defaults to `Levenshtein.ratio()`.
            dask_client: client
                A dask client object.
        """
        self.left = left
        self.right = right
        self.left_on = left_on
        self.right_on = right_on
        # Other attributes by default
        self.uselower = True
        self.threshold = 0.9
        self.how = 'outer'
        self.on = None
        self.left_index = False
        self.right_index = False
        self.parallel = True
        self.n_threads = 'all'
        self.hide_progress = False
        self.ratio_function = levenshtein_ratio
        self.dask_client = None

        # Get number of batches automatically
        # Define the batch size in number of rows
        batch_size = suggest_batch_number(
            self.left, self.right, print_enabled=False)
        # Get the total number of rows in the DataFrame
        total_rows = len(self.left)
        # Calculate the number of batches needed
        self.num_batches = (total_rows + batch_size - 1) // batch_size

    def set_parameter(self, parameter: str, value: Any):
        """Set a parameter to the `FuzzyMergeParallel` class.

        Args:
            parameter (str): Name of the parameter.
            value (any): Value to assign to the parameter
        """
        setattr(self, parameter, value)

    def __repr__(self):
        """Print a summary of the FuzzyMergeParallel object."""
        attributes = vars(self)
        attribute_strings = []

        for key, value in attributes.items():
            if isinstance(value, pd.DataFrame):
                attribute_strings.append(
                    f"{key} = {value.shape} - {value.columns}")
            else:
                attribute_strings.append(f"{key} = {value}")

        return "\n".join(attribute_strings)

    def merge(self):
        """Run the Fuzzy merge process and returns a data frame with merged dataframe with the matches.

        Returns:
            pd.DataFrame: A pandas dataframe with the merge result.
        """
        # Variables setting and validation
        if self.left_index:
            self.left_on = self.left_index

        if self.right_index:
            self.right_on = self.right_index

        if self.on:
            self.left_on = self.on
            self.right_on = self.on

        merge_on, delete_on = 'leftind', 'rightind'
        if self.how in ('left', 'inner'):
            merge_on, delete_on = 'leftind', 'rightind'
        elif self.how in ('right', 'outer'):
            merge_on, delete_on = 'rightind', 'leftind'
        else:
            raise ValueError(
                f"Parameter how not valid: {self.how}. Must be: left, inner, right, or outer.")

        if self.left_on is not None and self.left_on not in self.left.columns:
            raise KeyError(
                f"Column '{self.left_on}' does not exist in the left DataFrame")

        if self.right_on is not None and self.right_on not in self.right.columns:
            raise KeyError(
                f"Column '{self.right_on}' does not exist in the right DataFrame")

        left_tmp, right_tmp = self._get_np_subset()

        matched = self._match_by_left(left_tmp, right_tmp)

        mergeleft = self.left.merge(
            matched, how=self.how, left_index=True, right_on='leftind')

        mergeright = matched.merge(
            self.right, how=self.how, left_on='rightind', right_index=True)

        # Merge with simplified code
        merged = mergeleft.drop(columns=[delete_on]).merge(
            mergeright, how=self.how, on=merge_on).drop_duplicates(
            subset=[self.left_on, self.right_on]).drop(columns=['leftind', 'rightind'])

        if self.how in ['left', 'right', 'inner']:
            merged = merged.dropna(subset=[self.left_on] if self.how == 'left' else [self.right_on])
        elif self.how == 'outer':
            merged = merged.dropna(subset=[self.left_on, self.right_on], how='all')

        merged = merged.reset_index(drop=True)
        return merged

    def _get_np_subset(self):
        """Process the input dataframes to transform them into Numpy arrays with two columns.

            The first column is the index, the second column is the column of interest which is a string.

        Returns:
            Two numpy arrays
        """
        self.left = self.left.reset_index(drop=True)
        self.right = self.right.reset_index(drop=True)
        left_tmp = self.left.reset_index()[['index', self.left_on]].dropna(
            subset=[self.left_on])
        right_tmp = self.right.reset_index(
        )[['index', self.right_on]].dropna(subset=[self.right_on])

        if self.uselower:
            left_tmp[self.left_on] = left_tmp[self.left_on].str.lower()
            right_tmp[self.right_on] = right_tmp[self.right_on].str.lower()

        return left_tmp, right_tmp

    def _match_by_left(self, left_tmp: pd.DataFrame, right_tmp: pd.DataFrame) -> pd.DataFrame:
        """Apply a distance function between two dataframes that have a column of strings and returns a dataframe with the indices where the strings matched.

            Performs the following for each pair of strings of the two dataframes:
            1. Gets ratio between each pair of strings.
            2. Filters matches by the threshold.
            3. Returns a dataframe with the indices where the strings match according to the threshold.

            This function can run sequentially or in parallel by using multiprocessing.

        Args:
            left_tmp (pd.DataFrame): Dataframeof two columns. First column is an integer index. Second column is of strings (object).
            right_tmp (pd.DataFrame): Dataframeof two columns. First column is an integer index. Second column is of strings (object).

        Returns:
            pd.DataFrame: A two column dataframe containing the indices of the strings that match.
            First column are indices of left data, second column are indices of right data.
        """
        # Convert right dataframe to Numpy
        right_np = right_tmp[['index', self.right_on]].to_numpy()

        if self.parallel and self.dask_client is not None:  # Dask
            try:
                from fuzzymerge_parallel.matches_dask import match_by_left_dask
            except ImportError:
                import warnings
                warnings.warn(
                    "Dask dependencies not found, Please install 'fuzzymerge_parallel[dask]' to use this command.")
            else:
                matched = match_by_left_dask(left_tmp, right_np, self.left_on, self.threshold, self.dask_client, self.ratio_function, self.hide_progress)
                return matched

        # Convert left dataframe to Numpy
        left_np = left_tmp[['index', self.left_on]].to_numpy()

        print('Total batches:', self.num_batches,
              '| Rows per batch (approx):', left_np.shape[0] // self.num_batches)

        left_batches = np.array_split(left_np, self.num_batches)

        if self.parallel:  # multiprocessing
            matched = match_by_left_multiproc(left_batches, right_np, self.threshold, self.n_threads, self.ratio_function, self.hide_progress)
        else:  # Sequential
            matched = self._match_by_left_sequential(left_batches, right_np)
        return matched

    def _get_ratio_matrix(self, left_np: np.ndarray, right_np: np.ndarray) -> np.ndarray:
        """Apply a function to get the distance between two strings of each pair of two arrays.

        Args:
            left_np (np.ndarray): A one column array of strings.
            right_np (np.ndarray): A one column array of strings.

        Returns:
            np.ndarray: A Matrix with the difference measure between each pair of strings.
        """
        # Get rows and cols of matrix, create empty matrix to store results
        num_rows = left_np.shape[0]
        num_cols = right_np.shape[0]
        ratio_matrix = np.empty((num_rows, num_cols))
        for i in range(num_rows):
            ratio_matrix[i] = np.vectorize(
                self.ratio_function)(left_np[i], right_np)
        return ratio_matrix

    def _match_by_left_sequential(self, left_batches: list, right_np: np.ndarray) -> pd.DataFrame:
        """Sequential implementation of the function :func:`_match_by_left`."""
        # Iterate over each row pair and get Levenshtein ratio
        # for left_row in left_np:
        #    get_matches_left_row_mp(right_np, threshold, left_row)
        matched_batch_list = []
        n_matches = 0
        for left_batch in tqdm(left_batches, disable=self.hide_progress, desc=f"Matches {n_matches}"):
            # Get ratio matrix betwee left batch and right array
            ratio_matrix = self._get_ratio_matrix(
                left_batch[:, 1], right_np[:, 1])
            # Obtain the indexes where the ratio is >= than the threshold
            left_indices, right_indices = np.where(
                ratio_matrix >= self.threshold)
            # Get correct indices and convert to dataframe
            matched_batch = pd.DataFrame(
                {'leftind': left_batch[left_indices, 0], 'rightind': right_np[right_indices, 0]}, dtype=int)
            # Accumulate number of matches to print on progress bar
            n_matches += matched_batch.shape[0]
            matched_batch_list.append(matched_batch)

        matched = pd.concat(matched_batch_list)

        return matched
